### PR TITLE
Fog of war

### PR DESCRIPTION
**Pull request by OliverCartz**
Branch related to #68

### CODE REVIEW

File: aimmo-game/simulation/world_map.py [hunk 1]
```diff
@@ -12,6 +12,9 @@
 TARGET_NUM_PICKUPS_PER_AVATAR = 0.5
 PICKUP_SPAWN_CHANCE = 0.02
 
+NO_FOG_OF_WAR_DISTIANCE = 2
```
Comment (CelineBoudier) [inline aimmo-game/simulation/world_map.py]: not distance?

File: aimmo-game/simulation/avatar/fog_of_war.py [hunk 2]
```diff
@@ -7,4 +11,39 @@ def apply_fog_of_war(world_map, avatar_wrapper):
...
+def should_partially_fog(no_fog_distance, partial_fog_distance, x_dist, y_dist):
+    return x_dist > no_fog_distance or y_dist > no_fog_distance;
+
+def partially_fog_cell(cell):
+    partially_fogged_cell = copy.deepcopy(cell)
+    partially_fogged_cell.habitable = True
+    partially_fogged_cell.avatar = None
+    partially_fogged_cell.pickup = None
+    partially_fogged_cell.partially_fogged = True
+    return partially_fogged_cell;
```
Comment (Spycho) [inline aimmo-game/simulation/avatar/fog_of_war.py]: New line at EOF for PEP8 please

**Comment by Spycho** [comment]
After discussion, we thought to go with missing attributes, rather than flagging what data is "known", or having an "unknown" constant.

**Comment by joshuablake** [comment]
b227a2a has an alternate implementation which could allow removing `partially_fogged` from the API to address @mikebryant's concern. Also provides an easy way to add stuff to API with reduced duplication but more 'magic'.

File: aimmo-game-worker/simulation/world_map.py [hunk 3]
```diff
@@ -67,7 +72,7 @@ def can_move_to(self, target_location):
             cell = self.get_cell(target_location)
         except KeyError:
             return False
-        return cell.habitable and not cell.avatar
+        return getattr(cell, 'habitable', False) and not getattr(cell, 'avatar', False)
```
Comment (joshuablake) [inline aimmo-game-worker/simulation/world_map.py]: This method is now a bit non-intuitive to me. "Unknown" (ie: a fogged cell) is returned as False. Proposed alternative behaviour (using None as an unknown flag):
* For unfogged `cell.habitable and not cell.avatar` (previous behaviour).
* For partially fogged: `None if cell.habitable else False`
* For fully fogged: `None`.

Maybe check if it's on the map for fogged cells (if we're giving avatars that info)?

File: aimmo-game-worker/simulation/world_map.py [hunk 4]
```diff
@@ -51,7 +53,10 @@ def score_cells(self):
...
 
     def pickup_cells(self):
-        return (c for c in self.all_cells() if c.pickup)
+        return (c for c in self.all_cells() if getattr(self, 'pickup', False))
+
+    def partially_fogged_cells(self):
+        return (c for c in self.all_cells() if c.partially_fogged)
 
     def is_on_map(self, location):
         return location in self.cells
```
Comment (joshuablake) [inline aimmo-game-worker/simulation/world_map.py]: This now checks if the cell is in known/partially fogged (ie fully fogged cells give False). If we're not revealing map size then we should rename this method.

**Pull request merged**